odl/solvers/advanced/chambolle_pock.py

```diff
@@ -0,0 +1,278 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""First-order primal-dual algorithm developed by Chambolle and Pock."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+# External
+
+# Internal
+from odl.operator.operator import Operator
+from odl.operator.pspace_ops import ProductSpaceOperator
+
+__all__ = ('chambolle_pock_solver', 'f_cc_prox_l2_tv', 'g_prox_none')
+
+
+# TODO: add dual gap as convergence measure
+# TODO: diagonal preconditioning
+# TODO: add positivity constraint
+
+def chambolle_pock_solver(K, f_cc_prox, g_prox, sigma, tau, theta=1,
+                          niter=100, partial=None):
+    """Chambolle-Pock aglgorithms for convex optimization problems.
+
+    First order primal-dual hybrid-gradient (PDHG) method for non-smooth
+    convex optimization probelms with known saddle-point structure developed
+    by `Chambolle and Pock`_ (CP).
+
+
+    Parameters
+    ----------
+    K : `Operator` or `ProductSpaceOperator`
+    f_cc_prox : `function`
+       Factory function which, evaluated at ``sigma``, returns the proximal
+       operator of F^*. Domain and range of the returned proximal operator
+       is the range of ``K``.
+    g_prox : `function`
+        Factory function which, evaluated at ``tau``, returns the proximal
+       operator of ``G``. Domain and range of the returned proximal operator
+       is the domain of ``K``.
+    sigma : positive `float`
+        Convergence proofed for ``||K||_2^2 * sigma * tau < 1``
+    tau : positive `float`
+        Convergence proofed for ``||K||_2^2 * sigma * tau < 1``
+    theta : `float` in [0, 1], optional
+        Relaxation paramter
+    niter : `int`, optional
+        Number of iterations
+    partial : `Partial`, optional
+        If not `None` the object passed executes code per iteration,
+        e.g. plotting each iterate
+
+    Returns
+    -------
+    x : `DiscreteLpVector`
```

Comment:
I'll fix the pattern. The doc is actually wrong in the case `K` is a product space operator.

According to Chambolle and Pock the requirement actually is that `X` and `Y` are finite-dimensional Hilbert spaces given a mapping `K : X -> Y`. Since the default of `DiscreteLp` is L2 it's kind of correct. I will make a note in the docstring (which anyway needs some more words on the generic algorithms).